---
title: "visualize_16s"
author: "Reilly Cooper"
date: "9/30/2020"
output: html_document
---

```{r setup, include=FALSE}
library(here)
library(speedyseq)
library(ggpubr)
library(tidyverse)
library(cowplot)
library(rstatix)
library(vegan)
library(stringi)
library(DivNet)
library(parallel)
library(microbiome)
library(gt)

theme_set(theme_cowplot(font_size = 12))

phylum_colors <- c("Actinobacteriota" = "#B62A3D",
                   "Bacteroidota" = "#A35E60",
                   "Deinococcota" = "#456355",
                   "Dependentiae" = "#27223C",
                   "Desulfobacterota" = "#CC8B3C",
                   "Firmicutes" = "#EDCB64",
                   "Firmicutes_A" = "#EDCB65",
                   "Firmicutes_C" = "#EDCB66",
                   "Patescibacteria" = "#55506A",
                   "Planctomycetota" = "#6C6781",
                   "Proteobacteria" = "#003f5c",
                   "Verrucomicrobiota" = "#E6A2C5",
                   "Myxococcota" = "#2f4b7c",
                   "<=1% Abundant/Unidentified" = "#C4CFD0")

family_colors <- c("<=5% Abundant/Unidentified" = "#C4CFD0",
                   "Burkholderiaceae" = "#27223C",
                   "Chitinophagaceae" = "#A35E60",
                   "Flavobacteriaceae" = "#7496D2",
                   "Nanopelagicaceae" = "#55506A",
                   "Spirosomaceae" = "#E6A2C5",
                   "Vicingaceae" = "#76A08A")

genus_colors <- c("<=1% Abundant/Unidentified" = "#C4CFD0",
                  "<=5% Abundant/Unidentified" = "#C4CFD0",
                  "Acidovorax_D" = "#A35E60",
                  "Aquabacterium" = "#F0D9EC",
                  "Arcticibacterium" = "#456355",
                  "BRH-c54" = "#76A08A",
                  "Cypionkella" = "#F3E0E4",
                  "Emticicia" = "#CC8B3C",
                  "Flavobacterium" = "#7496D2",
                  "Fluviicola" = "#55506A",
                  "HdN1" = "#A4D5C3",
                  "Hydromonas" = "#6C6781",
                  "Hylemonella" = "#003f5c",
                  "Lacibacter" = "#E6A2C5",
                  "Limnobacter" = "#2f4b7c",
                  "Limnohabitans" = "#27223C",
                  "Methylophilus" = "#a1cacc",
                  "Methylotenera" = "#364362",
                  "Nevskia" = "#EDCB64",
                  "OLB11" = "#115189",
                  "Planktophila" = "#541F12",
                  "Polynucleobacter" = "#E8D2B9",
                  "Rubritalea" = "#B3ED99",
                  "Runella" = "#EAFDB4",
                  "Vitreoscilla_A" = "#AEA8A8")

id_colors <- c("day_dentifera" = "#76A08A",
               "night_dentifera" = "#27223C",
               "day_magna" = "#E6A2C5",
               "night_magna" = "#A35E60")

species_colors <- c("dentifera" = "#76A08A",
                    "magna" = "#E6A2C5")
```

Load data and remove Cyanobacteria reads, add clarity to Sample variable and add ID variable
```{r dataload}
data <- readRDS(here("data", "processed_16s_all.rds")) %>% 
  subset_taxa(Phylum != "Cyanobacteria")

data@sam_data$id <- c(rep("day_magna", 10), rep("night_magna", 10),
                                  rep("night_dentifera", 9), rep("day_dentifera", 10))

data@sam_data$Sample <- str_replace_all(data@sam_data$Sample, "Plate", "")

data@sam_data$Sample <- str_replace_all(data@sam_data$Sample , 
                                 c("13A01" = "DM_D1","13A02" = "DM_D2","13A03" = "DM_D3",
                                   "13A04" = "DM_D4","13A05" = "DM_D5","13A06" = "DM_D6",
                                   "13A07" = "DM_D7","13A08" = "DM_D8","13A09" = "DM_D9",
                                   "13A10" = "DM_D10","13B01" = "DM_N1","13B02" = "DM_N2",
                                   "13B03" = "DM_N3","13B04" = "DM_N4","13B05" = "DM_N5",
                                   "13B06" = "DM_N6","13B07" = "DM_N7","13B08" = "DM_N8",
                                   "13B09" = "DM_N9","13B10" = "DM_N10","13D01" = "DD_D1",
                                   "13D02" = "DD_D2","13D03" = "DD_D3","13D04" = "DD_D4",
                                   "13D05" = "DD_D5","13D06" = "DD_D6","13D07" = "DD_D7",
                                   "13D08" = "DD_D8","13D09" = "DD_D9","13D10" = "DD_D10",
                                   "13C01" = "DD_N1","13C02" = "DD_N2","13C03" = "DD_N3",
                                   "13C04" = "DD_N4","13C05" = "DD_N5","13C06" = "DD_N6",
                                   "13C07" = "DD_N7","13C08" = "DD_N8","13C09" = "DD_N9"))
```

Running PCoA on each of the four sample sets individually to check for outlier samples based on beta diversity.
```{r outlier_check}
magna_day <- subset_samples(data, id == "day_magna")
magna_night <- subset_samples(data, id == "night_magna")
dentifera_day <- subset_samples(data, id == "day_dentifera")
dentifera_night <- subset_samples(data, id == "night_dentifera")

mdord <- ordinate(magna_day, "PCoA", "unifrac", weighted = TRUE)
mnord <- ordinate(magna_night, "PCoA", "unifrac", weighted = TRUE)
ddord <- ordinate(dentifera_day, "PCoA", "unifrac", weighted = TRUE)
dnord <- ordinate(dentifera_night, "PCoA", "unifrac", weighted = TRUE)

mdplot <- plot_ordination(magna_day, mdord, color = "Sample") +
  geom_point(size = 2)
mnplot <- plot_ordination(magna_night, mnord, color = "Sample") +
  geom_point(size = 2)
ddplot <- plot_ordination(dentifera_day, ddord, color = "Sample") +
  geom_point(size = 2)
dnplot <- plot_ordination(dentifera_night, dnord, color = "Sample") +
  geom_point(size = 2)

ord_plots <- plot_grid(mdplot, mnplot,
          ddplot, dnplot,
          ncol = 2, nrow = 2, 
          labels = "auto")

ggsave(here("figures", "supplementary_figure_one.pdf"), ord_plots, units = "in", width = 7.5, height = 7.5, dpi = 300, useDingbats = FALSE)
```

Sample DD_D10 is much different than the rest of the _Daphnia dentifera_ day samples and the samples in general, so we are considering it an outlier. Even though it may not look like it at first glance, the X-axis is double the width of the other sample sets and this sample is significantly far to the right (0.1, compared to every other sample at 0.05 or less). Therefore, we're removing it from the dataset for subsequent analyses.
```{r}
data <- subset_samples(data, Sample != "DD_D10")
```


Creating a melted dataframe of the data, and reducing low abundance reads for clarity.
```{r melt}
data_transformed <- transform_sample_counts(data, function(x) x/sum(x))

melted <- psmelt(data_transformed) %>%
  select(-Sample) 

colnames(melted)[colnames(melted) == "sample_Sample"] <- "Sample"

no_low_abundance <- melted
no_low_abundance$Phylum[no_low_abundance$Abundance <= 0.01] <- "<=1% Abundant/Unidentified"
no_low_abundance$Phylum[is.na(no_low_abundance$Phylum)] <- "<=1% Abundant/Unidentified"
no_low_abundance$Family[no_low_abundance$Abundance <= 0.05] <- "<=5% Abundant/Unidentified"
no_low_abundance$Family[is.na(no_low_abundance$Family)] <- "<=5% Abundant/Unidentified"
```


Checking alpha diversity using DivNet instead of standard phyloseq alpha diversity measures. DivNet provides more accurate measures of diversity by accounting for unobserved taxa and random sampling from a larger community.
```{r divnet_alphadiv}
dv_data <- divnet(data, ncores = 4)

dv_data$shannon %>%
  summary %>%
  add_column("Sample" = data %>% sample_data %>% get_variable("Sample"))
  
dv_data$simpson %>%
  summary %>%
  add_column("Sample" = data %>% sample_data %>% get_variable("Sample"))

data_shannon <- dv_data$shannon %>% summary %>% select(sample_names, estimate) %>%
  mutate(id = data@sam_data$id) %>%
  mutate(species = data@sam_data$species) %>%
  mutate(time = data@sam_data$time)
data_simpson <- dv_data$simpson %>% summary %>% select(sample_names, estimate) %>%
  mutate(id = data@sam_data$id) %>%
  mutate(species = data@sam_data$species) %>%
  mutate(time = data@sam_data$time) %>%
  mutate(invsimpson = 1/estimate)

data_shannon$id <- factor(data_shannon$id, 
                          levels = c("day_dentifera", 
                                     "night_dentifera", 
                                     "day_magna", 
                                     "night_magna"))

data_simpson$id <- factor(data_shannon$id, 
                          levels = c("day_dentifera", 
                                     "night_dentifera", 
                                     "day_magna", 
                                     "night_magna"))

magnasimp <- data_simpson %>% filter(species == "magna") %>% mutate(invsimp = 1/estimate)
dentsimp <- data_simpson %>% filter(species == "dentifera") %>% mutate(invsimp = 1/estimate)
```

Figure 1.
Differences between species, regardless of day or night. This is a three panel figure:
a) Alpha diversity measures
b) Genus-level composition
d) Beta diversity measure, weighted UniFrac PCoA
```{r mag_v_dent}
species_comp_data <- merge_samples2(data, group = "species")

scdt <- transform_sample_counts(species_comp_data, function(x) x/sum(x))

species_melted <- psmelt(scdt) %>%
  select(-Sample, replicate, X)

colnames(species_melted)[colnames(species_melted) == "sample_Sample"] <- "Sample"

species_phylumsummary <- species_melted %>% group_by(Phylum) %>% summarise(phylum_abundance = sum(Abundance)) %>%
  mutate(new_phylum = ifelse(phylum_abundance < 0.01, "<=1% Abundant/Unidentified", Phylum))
species_familysummary <- species_melted %>% group_by(Family) %>% summarise(family_abundance = sum(Abundance)) %>%
  mutate(new_family = ifelse(family_abundance < 0.05, "<=5% Abundant/Unidentified", Family))
species_genussummary <- species_melted %>% group_by(Genus) %>% summarise(genus_abundance = sum(Abundance)) %>%
  mutate(new_genus = ifelse(genus_abundance < 0.01, "<=1% Abundant/Unidentified", Genus))

species_melted <- merge(species_melted, species_phylumsummary, by = "Phylum", all.x = TRUE) %>%
  mutate(new_phylum = ifelse(is.na(new_phylum), "<=1% Abundant/Unidentified", new_phylum))
species_melted <- merge(species_melted, species_familysummary, by = "Family", all.x = TRUE) %>%
  mutate(new_family = ifelse(is.na(new_phylum), "<=5% Abundant/Unidentified", new_family))
species_melted <- merge(species_melted, species_genussummary, by = "Genus", all.x = TRUE) %>%
  mutate(new_genus = ifelse(is.na(new_genus), "<=1% Abundant/Unidentified", new_genus))

species_melted$species <- factor(species_melted$species, levels = c("magna", "dentifera"))
data_simpson$species <- factor(data_simpson$species, levels = c("magna", "dentifera"))

species_alphadiv_plot <- ggplot(data_simpson, aes(species, 1/estimate, group = species, color = species)) + 
  geom_violin(lwd = 1, width = .5) + 
  geom_jitter(height = 0, width = 0.05, alpha = 0.5, size = 2) +
  scale_color_manual(values = species_colors) +
  ylab("Inverse Simpson Index Estimate") +
  theme(axis.title.x = element_blank(),
        legend.position = "none") +
  scale_x_discrete(labels = c("Daphnia\nmagna",
                              "Daphnia\ndentifera")) +
  stat_compare_means(method = "t.test", label.y = 15, label.x = 0.75, label = "p.format")

species_composition_plot <- ggplot(species_melted, aes(species, Abundance, fill = new_genus)) +
  geom_bar(stat = "identity", position = "stack") +
  scale_fill_manual(values = genus_colors) +
  theme(axis.title.x = element_blank()) +
  ylab("Relative abundance") +
  scale_x_discrete(labels = c("Daphnia\nmagna",
                              "Daphnia\ndentifera"))

species_composition_plot
wuniord <- ordinate(data, "PCoA", "unifrac", weighted = TRUE)
data.rel <- microbiome::transform(data, "compositional")
data.otu <- microbiome::abundances(data.rel)
data.meta <- microbiome::meta(data.rel)
data.unifrac <- UniFrac(data.rel,
                        weighted = TRUE,
                        normalized = TRUE,
                        parallel = FALSE,
                        fast = TRUE)
data.permanova <- adonis(data.unifrac ~ species, data = data.meta)

label <- paste("~italic(R)^2==~", round(data.permanova$aov.tab$R2[1], digits = 3))
p <- paste("~italic(P)==~", round(data.permanova$aov.tab$`Pr(>F)`[1], digits = 3))

w_ord_plot <- plot_ordination(data, wuniord, color = "species") +
  geom_point(size = 2) +
  scale_color_manual(values = species_colors) +
  stat_ellipse(linetype = 2) +
  annotate("text", x = 0, y = -0.075, label = label, parse = TRUE, size = 3.5) +
  annotate("text", x = 0, y = -0.085, label = p, parse = TRUE, size = 3.5)


figure_one <- plot_grid(plot_grid(species_alphadiv_plot, 
                                  species_composition_plot, 
                                  nrow = 1, 
                                  labels = c("a", "b"),
                                  rel_widths = c(0.5, .75), align = "h"),
                        plot_grid(NULL, 
                                  w_ord_plot, 
                                  NULL, 
                                  nrow = 1, 
                                  rel_widths = c(.25, .5, .25), 
                                  labels = c("", "c", "")), 
                        nrow = 2, 
                        align = "hv")

ggsave(here("figures", "figure_one.pdf"), 
       figure_one, 
       units = "in", 
       width = 8, 
       height = 8, 
       dpi = 300, 
       useDingbats = FALSE)

# Alpha diversity statistics (t-test) for host species
data_simpson <- data_simpson %>% mutate(invsimpson = 1/estimate)
data_simpson %>% t_test(invsimpson ~ species) %>% write_csv(here("statistics", "species_alphadiv_ttest.csv"))

# Beta diversity statistics (PERMANOVA) for host species
data.permanova
```

Day vs. night differences in both Daphnia species. 
```{r day_night, warning=FALSE}
data_by_sp_time <- data
transformed_sp_time <- transform_sample_counts(data_by_sp_time, function(x) x/sum(x))

sp_time_melted <- psmelt(transformed_sp_time) %>%
  select(-c(replicate, X))

species_phylumsummary <- sp_time_melted %>% group_by(Phylum) %>% summarise(phylum_abundance = sum(Abundance)) %>%
  mutate(new_phylum = ifelse(phylum_abundance < 0.01, "<=1% Abundant/Unidentified", Phylum))
species_familysummary <- sp_time_melted %>% group_by(Family) %>% summarise(family_abundance = sum(Abundance)) %>%
  mutate(new_family = ifelse(family_abundance < 0.05, "<=5% Abundant/Unidentified", Family))
species_genussummary <- sp_time_melted %>% group_by(Genus) %>% summarise(genus_abundance = sum(Abundance)) %>%
  mutate(new_genus = ifelse(genus_abundance < 0.05, "<=5% Abundant/Unidentified", Genus))

sp_time_melted <- merge(sp_time_melted, species_phylumsummary, by = "Phylum", all.x = TRUE) %>%
  mutate(new_phylum = ifelse(is.na(new_phylum), "<=1% Abundant/Unidentified", new_phylum))
sp_time_melted <- merge(sp_time_melted, species_familysummary, by = "Family", all.x = TRUE) %>%
  mutate(new_family = ifelse(is.na(new_phylum), "<=5% Abundant/Unidentified", new_family))
sp_time_melted <- merge(sp_time_melted, species_genussummary, by = "Genus", all.x = TRUE) %>%
  mutate(new_genus = ifelse(is.na(new_genus), "<=5% Abundant/Unidentified", new_genus))

# Alpha diversity

# ANOVA and t-tests for differences among species and times for alpha diversity
data_simpson <- data_simpson[-39,]
time_spp_aov <- aov(invsimpson ~ species*time, data = data_simpson)
summary(time_spp_aov)

time_spp_tukeyhsd <- time_spp_aov %>% tukey_hsd()
time_spp_tukeyhsd %>% write_csv(here("statistics", "species_time_tukeyhsd.csv"))
comparisons <- time_spp_tukeyhsd %>% 
  filter((group1 == "magna:day" & group2 == "magna:night") | (group1 == "dentifera:day" & group2 == "dentifera:night")) %>%
  mutate(group1 = gsub("magna:day", "day_magna", group1),
         group1 = gsub("dentifera:day", "day_dentifera", group1),
         group2 = gsub("magna:night", "night_magna", group2),
         group2 = gsub("dentifera:night", "night_dentifera", group2))

data_simpson$id <- factor(data_simpson$id, 
                          levels = c("day_magna", 
                                     "night_magna", 
                                     "day_dentifera", 
                                     "night_dentifera"))

time_alphadiv_plot <- ggplot(data_simpson, aes(id, 1/estimate, group = id, color = id)) + 
  geom_violin() +
  geom_jitter(height = 0, width = 0.05, alpha = 0.5, size = 2) +
  scale_color_manual(values = id_colors) +
  theme(axis.title.x = element_blank(),
        axis.text.x = element_text(size = 8),
        legend.position = "none") +
  scale_x_discrete(labels = c("Daphnia\nmagna,\nday",
                              "Daphnia\nmagna,\nnight",
                              "Daphnia\ndentifera,\nday",
                              "Daphnia\ndentifera,\nnight")) +
  annotate("text", x = 1, y = 15, label = "Anova, p < 2e-16") +
  stat_pvalue_manual(comparisons, label = "p.adj", y.position = c(18, 18)) +
  ylab("Inverse Simpson Index Estimate")

# Composition, genus level
sp_time_melted %>% mutate(identifier = ifelse(time == "day" & species == "magna", "Daphnia\nmagna,\nday", 
                                              ifelse(time == "day" & species == "dentifera", "Daphnia\ndentifera,\nday", 
                                                     ifelse(time == "night" & species == "magna", "Daphnia\nmagna,\nnight", "Daphnia\ndentifera,\nnight")))) -> sp_time_melted

sp_time_melted$identifier <- factor(sp_time_melted$identifier, levels = c("Daphnia\nmagna,\nday", "Daphnia\nmagna,\nnight",
                                                                          "Daphnia\ndentifera,\nday", "Daphnia\ndentifera,\nnight"))

time_composition_plot <- ggplot(sp_time_melted, aes(Sample, Abundance, fill = new_genus)) +
  geom_bar(stat = "identity", position = "stack") +
  scale_fill_manual(values = genus_colors) +
  theme(axis.title.x = element_blank(),
        axis.text.x = element_blank(),
        strip.background = element_blank(),
        strip.placement = "outside", 
        strip.text.x = element_text(size = 8),
        legend.text = element_text(size = 8),
        legend.title = element_text(size = 10),
        legend.key.size = unit(0.5, "line")) +
  ylab("Relative abundance") +
  facet_wrap(~identifier, scales = "free_x", ncol = 4, strip.position = "bottom") +
  guides(fill = guide_legend(ncol = 1)) + 
  labs(fill = "Genus")

figure_two <- plot_grid(time_alphadiv_plot, 
                    time_composition_plot, 
                    nrow = 1, rel_widths = c(0.5, 1),
                    align = "hv", labels = "auto")

ggsave(here("figures", "figure_two.pdf"), 
       figure_two, 
       units = "in", 
       width = 8, 
       height = 4, 
       dpi = 300, 
       useDingbats = FALSE)

# Will fix "Genus" and Anova value overlap in Illustrator
```


Identifying core microbiota between host species.
```{r core, warning=FALSE}
magna <- subset_samples(data, species == "magna")
dentifera <- subset_samples(data, species == "dentifera")

magna_rel <- microbiome::transform(magna, "compositional")
dentifera_rel <- microbiome::transform(dentifera, "compositional")

magna_nolow <- prune_taxa(taxa_sums(magna_rel) > 0, magna_rel)
dentifera_nolow <- prune_taxa(taxa_sums(dentifera_rel) > 0, dentifera_rel)

magna_core <- core(magna_nolow, detection = 0.001, prevalence = 50/100)
dentifera_core <- core(dentifera_nolow, detection = 0.001, prevalence = 50/100)

magna_core_df <- as.data.frame(magna_core@tax_table) %>% rownames_to_column("ASV")
dentifera_core_df <- as.data.frame(dentifera_core@tax_table) %>% rownames_to_column("ASV")

in_both <- merge(magna_core_df, dentifera_core_df) %>% mutate(core = "Both")
magna_only <- anti_join(magna_core_df, dentifera_core_df) %>% mutate(core = "Daphnia magna")
dentifera_only <- anti_join(dentifera_core_df, magna_core_df) %>% mutate(core = "Daphnia dentifera")

core_table <- do.call("rbind", list(in_both, magna_only, dentifera_only))

# Getting mean relative abundance of each ASV in core
asvs <- core_table$ASV

find_asv_df <- melted
colnames(find_asv_df)[1] <- "ASV"
core_filt <- filter(find_asv_df, ASV %in% asvs)

core_magna_abundance <- core_filt %>% filter(species == "magna") %>%
  group_by(ASV) %>%
  summarise(magna_abundance = mean(Abundance))
core_dentifera_abundance <- core_filt %>% filter(species == "dentifera") %>%
  group_by(ASV) %>%
  summarise(dentifera_abundance = mean(Abundance))

core_table <- merge(core_table, core_magna_abundance, all = TRUE) %>% merge(core_dentifera_abundance, all = TRUE)

core_table$core <- factor(core_table$core, levels = c("Both", "Daphnia magna", "Daphnia dentifera"))
# Actual ASVs go in supplemental table, for clarity

core_table_ordered <- core_table %>%
  mutate(core = fct_relevel(core, c("Both", "Daphnia magna", "Daphnia dentifera"))) %>%
  arrange(core, desc(magna_abundance), desc(dentifera_abundance)) %>%
  relocate(ASV, .after = dentifera_abundance) %>%
  relocate(c(core, magna_abundance, dentifera_abundance), .before = Kingdom) %>%
  mutate_if(is.numeric, round, 3) %>%
  select(-c(Kingdom, ASV)) %>%
  gt() %>%
  cols_label(magna_abundance = md("Mean relative abundance in\n_Daphnia magna_"),
             dentifera_abundance = md("Mean relative abundance in\n_Daphnia dentifera_"),
             core = "Core to:") %>%
  data_color(columns = vars(core),
             colors = scales::col_factor(palette = c("#27223C", "#76A08A", "#E6A2C5"),
                                         domain = c("Both", "Daphnia dentifera", "Daphnia magna")),
             alpha = 0.5) %>%
  data_color(columns = vars(magna_abundance),
             colors = scales::col_numeric(palette = c("#469990", "#93E6DD"),
                                          domain = c(0, 0.512)),
             alpha = 0.5) %>%
  data_color(columns = vars(dentifera_abundance),
             colors = scales::col_numeric(palette = c("#469990", "#93E6DD"),
                                          domain = c(0, 0.154)),
             alpha = 0.5) %>%
  tab_spanner(label = "Taxonomic identity",
              columns = vars(Class, Order, Family, Genus, Species)) %>%
  cols_align(align = "left") %>%
  tab_options(heading.align = "left",
              column_labels.border.bottom.width = px(2),
              column_labels.border.bottom.color = "black",
              table.border.bottom.color = "black",
              table.border.bottom.width = px(2))

gtsave(core_table_ordered, here("figures", "table_one.pdf"))

```

Table and figure 3: Taxa with abundance > 0.005 that are in only Daphnia dentifera OR only in Daphnia magna, OR both at time points
```{r}
merged_by_id <- merge_samples2(data, "id")

comp_id <- transform_sample_counts(merged_by_id, function(x) x/sum(x))

comp_id_melted <- psmelt(comp_id)

abund <- comp_id_melted %>% filter(Abundance > 0.005)

day_mag <- abund %>% filter(Sample == "day_magna") %>% select(OTU)
night_mag <- abund %>% filter(Sample == "night_magna") %>% select(OTU)
day_dent <- abund %>% filter(Sample == "day_dentifera") %>% select(OTU)
night_dent <- abund %>% filter(Sample == "night_dentifera") %>% select(OTU)

magna <- rbind(day_mag, night_mag)
dentifera <- rbind(day_dent, night_dent)

magna_otus <- magna %>% select(OTU) %>% unique()
dentifera_otus <- dentifera %>% select(OTU) %>% unique()

magna_only <- setdiff(magna_otus, dentifera_otus)
dentifera_only <- setdiff(dentifera_otus, magna_otus)

magna_only <- magna_only %>% mutate(location = "magna")
magna_otus <- merge(magna_otus, magna_only, by = "OTU", all.x = TRUE) %>% 
  mutate(location = ifelse(is.na(location), "both", location))

dentifera_only <- dentifera_only %>% mutate(location = "dentifera")
dentifera_otus <- merge(dentifera_otus, dentifera_only, by = "OTU")

all_otus <- rbind(magna_otus, dentifera_otus)

abund <- merge(abund, all_otus, by = "OTU", all.x = TRUE)

important_otus <- melted %>% filter(OTU %in% abund$OTU)

important_otus <- merge(important_otus, all_otus, by = "OTU", all.x = TRUE)

# Adding ASV identifiers for ASVs in genera present in both, but specific ASV only present in one or the other
important_otus <- important_otus %>% mutate(Genus = ifelse(is.na(Genus), Family, Genus)) %>%
  mutate(Genus = ifelse(is.na(Genus), Order, Genus)) %>%
  mutate(plotted_genus = Genus) %>%
  mutate(plotted_genus = ifelse(plotted_genus == "Limnohabitans" & location == "dentifera", paste0(plotted_genus, " ASV 1"), plotted_genus)) %>%
  mutate(plotted_genus = ifelse(plotted_genus == "Limnohabitans" & location == "magna", paste0(plotted_genus, " ASV 2"), plotted_genus)) %>%
  mutate(plotted_genus = ifelse(plotted_genus == "Limnohabitans" & location == "both", paste0(plotted_genus, " ASV 3"), plotted_genus)) %>%
  mutate(plotted_genus = ifelse(plotted_genus == "Burkholderiaceae" & location == "dentifera", paste0(plotted_genus, " ASV 1"), plotted_genus)) %>%
  mutate(plotted_genus = ifelse(plotted_genus == "Burkholderiaceae" & location == "magna", paste0(plotted_genus, " ASV 2"), plotted_genus)) %>%
  mutate(plotted_genus = ifelse(plotted_genus == "Emticicia" & location == "magna", paste0(plotted_genus, " ASV 1"), plotted_genus)) %>%
  mutate(plotted_genus = ifelse(plotted_genus == "Emticicia" & location == "both", paste0(plotted_genus, " ASV 2"), plotted_genus))

dent_otus_allsamples <- important_otus %>% filter(location == "dentifera") %>% filter(species == "dentifera")
mag_otus_allsamples <- important_otus %>% filter(location == "magna") %>% filter(species == "magna")

dent_otus_meanabund <- dent_otus_allsamples %>% group_by(plotted_genus) %>% summarise(mean_abund = mean(Abundance)) %>% arrange(mean_abund)
dent_levels <- as.list(dent_otus_meanabund$plotted_genus)
dent_otus_allsamples$plotted_genus <- factor(dent_otus_allsamples$plotted_genus, levels = dent_levels)

mag_otus_meanabund <- mag_otus_allsamples %>% group_by(plotted_genus) %>% summarise(mean_abund = mean(Abundance)) %>% arrange(mean_abund)
mag_levels <- as.list(mag_otus_meanabund$plotted_genus)
mag_otus_allsamples$plotted_genus <- factor(mag_otus_allsamples$plotted_genus, levels = mag_levels)

dent_only_plot <- ggplot(dent_otus_allsamples, aes(plotted_genus, Abundance, color = id)) + 
  geom_boxplot(outlier.shape = NA) + 
  geom_jitter(alpha = 0.5, position = position_jitterdodge()) + 
  scale_color_manual(values = id_colors, 
                     labels = c("Daphnia dentifera,\nday",
                                "Daphnia dentifera,\nnight"),
                     name = "") +
  theme(axis.title.y = element_blank(),
        strip.background = element_blank(),
        strip.text.x = element_text(size = 8), 
        legend.position = "none") +
  ylab("Genus Relative Abundance") +
  stat_compare_means(method = "t.test", aes(label = ..p.signif..)) +
  coord_flip()

mag_only_plot <- ggplot(mag_otus_allsamples, aes(plotted_genus, Abundance, color = id)) + 
  geom_boxplot(outlier.shape = NA) + 
  geom_jitter(alpha = 0.5, position = position_jitterdodge()) + 
  scale_color_manual(values = id_colors, 
                     labels = c("Daphnia magna,\nday",
                                "Daphnia magna,\nnight"),
                     name = "") +
  theme(axis.title.y = element_blank(),
        strip.background = element_blank(),
        strip.text.x = element_text(size = 8), 
        legend.position = "none") +
  ylab("Genus Relative Abundance") +
  stat_compare_means(method = "t.test", aes(label = ..p.signif..)) +
  coord_flip()

both_otus_allsamples <- important_otus %>% filter(location == "both")
both_otus_meanabund <- both_otus_allsamples %>% group_by(plotted_genus) %>% summarise(mean_abund = mean(Abundance)) %>% arrange(mean_abund)
both_levels <- as.list(both_otus_meanabund$plotted_genus)
both_otus_allsamples$plotted_genus <- factor(both_otus_allsamples$plotted_genus, levels = both_levels)

both_only_plot <- ggplot(both_otus_allsamples, aes(plotted_genus, Abundance, color = id)) + 
  geom_boxplot(outlier.shape = NA) + 
  geom_jitter(alpha = 0.5, position = position_jitterdodge()) + 
  scale_color_manual(values = id_colors, 
                     labels = c("Daphnia dentifera,\nday",
                                "Daphnia dentifera,\nnight",
                                "Daphnia magna,\nday",
                                "Daphnia magna,\nnight"),
                     name = "") +
  theme(axis.title.y = element_blank(),
        strip.background = element_blank(),
        strip.text.x = element_text(size = 8)) +
  ylab("Genus Relative Abundance") +
  stat_compare_means(method = "anova", aes(label = ..p.signif..)) +
  coord_flip()


species_specific <- plot_grid(mag_only_plot, dent_only_plot, labels = "auto", align = "h")
both_with_blankspace <- plot_grid(NULL, both_only_plot, NULL, rel_widths = c(.25, 1, .1), labels = c("", "c", ""), nrow = 1)

complete_asv_plot <- plot_grid(species_specific, both_with_blankspace, nrow = 2)

ggsave(here("figures", "figure_three.pdf"), 
       complete_asv_plot, 
       units = "in", 
       width = 8, 
       height = 8, 
       dpi = 300, 
       useDingbats = FALSE)

# TODO: Making output table
mag_otus_bygenus <- split(mag_otus_allsamples, f = mag_otus_allsamples$plotted_genus)

mag_ttests <- lapply(mag_otus_bygenus, function(df){
  res <- t_test(Abundance ~ id, data = df)
  res <- mutate(res, species = "magna")
  res
})

mag_ttest_results <- plyr::ldply(mag_ttests, data.frame)

dent_otus_bygenus <- split(dent_otus_allsamples, f = dent_otus_allsamples$plotted_genus)

dent_ttests <- lapply(dent_otus_bygenus, function(df){
  res <- t_test(Abundance ~ id, data = df)
  res <- mutate(res, species = "dentifera")
  res
})

dent_ttest_results <- plyr::ldply(dent_ttests, data.frame)



all_ttests <- rbind(mag_ttest_results, dent_ttest_results)
# Prettifying
pretty_all_ttests <- all_ttests %>% select(.id, statistic, df, p, species)
colnames(pretty_all_ttests) <- c("ASV", "statistic", "df", "p", "species")
pretty_all_ttests %>%
  gt(rowname_col = "ASV") %>%
  tab_row_group(
    group = "Daphnia magna",
    rows = species == "magna"
  ) %>%
  tab_row_group(
    group = "Daphnia dentifera",
    rows = species == "dentifera"
  ) %>%
  cols_hide(
    columns = vars(species)
  ) %>%
  cols_label(
    statistic = "Statistic",
    df = "Df",
    p = html("<i>p</i>")
  ) %>%
  gtsave(here("statistics", "bacteria_genus_ttests.pdf"))

# Anovas for ASVs found in both host species
both_otus_bygenus <- split(both_otus_allsamples, f = both_otus_allsamples$plotted_genus)

both_anovas <- lapply(both_otus_bygenus, function(df){
  anovares <- aov(Abundance ~ id, data = df)
  res <- anova_summary(anovares)
  res
})

both_aov_results <- plyr::ldply(both_anovas, data.frame)
# Prettifying
pretty_aov_results <- both_aov_results %>% select(.id, DFn, DFd, F, p, ges)
colnames(pretty_aov_results) <- c("ASV", "DFn", "DFd", "F-value", "p-value", "Effect Size")
pretty_aov_results %>%
  gt(rowname_col = "ASV") %>%
  cols_label(
    `p-value` = html("<i>p</i>-value")
  ) %>%
  gtsave(here("statistics", "bacteria_genus_anova.pdf"))

both_tukeys <- lapply(both_otus_bygenus, function(df) {
  tukeys <- aov(Abundance ~ id, data = df) %>% tukey_hsd()
  tukeys
})

both_tukey_results <- plyr::ldply(both_tukeys, data.frame)
pretty_tukey_results <- both_tukey_results %>% select(-term)
colnames(pretty_tukey_results) <- c("ASV", "Group 1", "Group 2", "Null value", "Estimate", "CI (low)", "CI (high)", "p-value (adj.)", "p-value (adj.) significance")
pretty_tukey_results %>%
  gt(rowname_col = "ASV") %>%
  cols_label(
    `p-value (adj.)` = html("<i>p</i>-value (adj.)")
  ) %>%
  gtsave(here("statistics", "bacteria_both_tukey.pdf"))
```

